### PR TITLE
Add support for restoring from snapshot with search replicas

### LINKED ISSUE #15532: [RW Separation] Updates to restore from snapshot with added search replicas
With https://github.com/opensearch-project/OpenSearch/issues/15368 we add the ability to create and update an index to have search replicas.  We need to also support this configuration on restore.  With this issue we should accomplish this while honoring the same validations for search replicas:
1. Search replicas allowed with Segrep only
2. Auto expand replica count not allowed
3. Feature flag enforced

### PR DESCRIPTION
**Pull request by vinaykpud**
### Description

In this PR we are adding,

1. Support for restoring search replicas from a snapshot taken with search replica
2. Add validations to ensure compatibility between `DOCUMENT` and `SEGMENT` replication type while snapshot restoration. Following are the allowed rules:
* Snapshots taken when the replication type is `DOCUMENT` can be restored to `DOCUMENT` or `SEGMENT` types. If we are restoring to `SEGMENT` type more search replicas can be added. 
* Snapshots taken when the replication type is `SEGMENT` and without any search replicas can be restored to `DOCUMENT` or `SEGMENT` types. If we are restoring to `SEGMENT` type more search replicas can be added. 
* Snapshots taken when the replication type is `SEGMENT` and with search replicas can be restored to `DOCUMENT` or `SEGMENT` types. If we are restoring to `SEGMENT` type more search replicas can be added. If we are restoring to `DOCUMENT` type search replica count must be set to 0 in the settings.

### Related Issues
Resolves #15532 
Related #15306 

### Check List
- [x] Functionality includes testing.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/snapshots/RestoreService.java [hunk 1]
```diff
@@ -650,6 +661,23 @@ public ClusterState execute(ClusterState currentState) {
                         return allocationService.reroute(updatedState, "restored snapshot [" + snapshot + "]");
                     }
 
+                    private void validateReplicationTypeRestoreSettings(String snapshotReplicationType,
```
Comment (mch2) [inline server/src/main/java/org/opensearch/snapshots/RestoreService.java:664]: nitpicks - 

Lets pass a ReplicationType here and parse in the caller and use `ReplicationType. DOCUMENT.equals(replicationType)`

is `snapshotIndexMetadata` the metadata inside the snapshot? Passing that as `restoreReplicationType` is a bit confusing.  How about `snapshottedMetadata` and `updatedMetadata` ?

File: server/src/test/java/org/opensearch/cluster/routing/SearchOnlyReplicaRestoreTests.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        RoutingTable routingTable = RoutingTable.builder()
+            .addAsNewRestore(indexMetadata, snapshotRecoverySource, new HashSet<>())
+            .build();
+
+        ClusterState clusterState = ClusterState.builder(ClusterName.CLUSTER_NAME_SETTING.getDefault(Settings.EMPTY))
+            .metadata(metadata)
+            .routingTable(routingTable)
+            .build();
+
+        List<ShardRouting> shardRoutings = clusterState.routingTable().index("test").shard(0).getShards();
```
Comment (mch2) [inline server/src/test/java/org/opensearch/cluster/routing/SearchOnlyReplicaRestoreTests.java:57]: `IndexShardRoutingTable` has a function `searchOnlyReplicas` that will return this list for you.

File: server/src/internalClusterTest/java/org/opensearch/indices/replication/SearchReplicaRestoreIT.java [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+            replicationType));
+        indexRandomDocs(INDEX_NAME, DOC_COUNT);
+        refresh(INDEX_NAME);
+        ensureGreen(INDEX_NAME);
+    }
+
+    private void startCluster(int numOfNodes) {
+        internalCluster().startClusterManagerOnlyNode();
+        for(int i = 0; i< numOfNodes; i++) {
+            internalCluster().startDataOnlyNode();
```
Comment (mch2) [inline server/src/internalClusterTest/java/org/opensearch/indices/replication/SearchReplicaRestoreIT.java:255]: There should already be a method `startDataOnlyNodes(int)` that you can use and the inline this into each bootstrap method

File: server/src/internalClusterTest/java/org/opensearch/indices/replication/SearchReplicaRestoreIT.java [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+        );
+        ensureGreen(RESTORED_INDEX_NAME);
+
+        SearchResponse resp = client().prepareSearch(RESTORED_INDEX_NAME).setQuery(QueryBuilders.matchAllQuery()).get();
+        assertHitCount(resp, DOC_COUNT);
+    }
+
+    private void bootstrapIndexWithOutSearchNodes(ReplicationType replicationType) throws InterruptedException {
+        startCluster(2);
+        createIndex(INDEX_NAME, getIndexSettings(1, 1, 0,
```
Comment (mch2) [inline server/src/internalClusterTest/java/org/opensearch/indices/replication/SearchReplicaRestoreIT.java:236]: nitpick - would rather just build the settings here inline for each case, its a lot easier to read imo

File: server/src/internalClusterTest/java/org/opensearch/indices/replication/SearchReplicaRestoreIT.java [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+            replicationType));
+        indexRandomDocs(INDEX_NAME, DOC_COUNT);
+        refresh(INDEX_NAME);
+        ensureGreen(INDEX_NAME);
+    }
+
+    private void bootstrapIndexWithSearchNodes(ReplicationType replicationType) throws InterruptedException {
```
Comment (mch2) [inline server/src/internalClusterTest/java/org/opensearch/indices/replication/SearchReplicaRestoreIT.java:243]: you can't boostrap with docrep here, so I don't think you need the param?

File: server/src/internalClusterTest/java/org/opensearch/indices/replication/SearchReplicaRestoreIT.java [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+        assertHitCount(resp, DOC_COUNT);
+    }
+
+    public void testSearchReplicaRestore_WhenSnapshotOnSegRepWithSearchReplica_RestoreOnSegRep() throws Exception {
+        bootstrapIndexWithSearchNodes(ReplicationType.SEGMENT);
+        createRepoAndSnapshot(REPOSITORY_NAME, FS_REPOSITORY_TYPE, SNAPSHOT_NAME, INDEX_NAME);
+
+        restoreSnapshot(REPOSITORY_NAME, SNAPSHOT_NAME, INDEX_NAME, RESTORED_INDEX_NAME,
+            Settings.builder()
+                .put(IndexMetadata.SETTING_REPLICATION_TYPE, ReplicationType.SEGMENT)
```
Comment (mch2) [inline server/src/internalClusterTest/java/org/opensearch/indices/replication/SearchReplicaRestoreIT.java:225]: i don't think you need to worry about explicitly setting the replication type unless you intend to change it.  Same for other tests

File: server/src/internalClusterTest/java/org/opensearch/indices/replication/SearchReplicaRestoreIT.java [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+                .build()
+        );
+        ensureGreen(RESTORED_INDEX_NAME);
```
Comment (mch2) [inline server/src/internalClusterTest/java/org/opensearch/indices/replication/SearchReplicaRestoreIT.java:228]: we should assert search shard count post restore.

File: server/src/internalClusterTest/java/org/opensearch/indices/replication/SearchReplicaRestoreIT.java [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+            .put(super.featureFlagSettings())
+            .put(FeatureFlags.READER_WRITER_SPLIT_EXPERIMENTAL, true)
+            .build();
+    }
+
+    public void testSearchReplicaRestore_WhenSnapshotOnSegRep_RestoredWithSameSettings() throws Exception {
+        bootstrapIndexWithOutSearchNodes(ReplicationType.SEGMENT);
+        createRepoAndSnapshot(REPOSITORY_NAME, FS_REPOSITORY_TYPE, SNAPSHOT_NAME, INDEX_NAME);
+
+        restoreSnapshot(REPOSITORY_NAME, SNAPSHOT_NAME, INDEX_NAME, RESTORED_INDEX_NAME, null);
```
Comment (mch2) [inline server/src/internalClusterTest/java/org/opensearch/indices/replication/SearchReplicaRestoreIT.java:48]: I don't think you need this test as you are testing base snapshot restore behavior that has nothing to do with search replicas?

File: server/src/test/java/org/opensearch/cluster/routing/SearchOnlyReplicaRestoreTests.java (hunk 2 above)
Comment (vinaykpud) [reply server/src/test/java/org/opensearch/cluster/routing/SearchOnlyReplicaRestoreTests.java:57]: Sure

File: server/src/main/java/org/opensearch/snapshots/RestoreService.java (hunk 1 above)
Comment (vinaykpud) [reply server/src/main/java/org/opensearch/snapshots/RestoreService.java:664]: `snapshotIndexMetadata` contains metadata of snapshot and whatever setting user is trying to override. So basically its a metadata which will be applied for resulting restored index.

Comment (vinaykpud) [reply server/src/main/java/org/opensearch/snapshots/RestoreService.java:664]: do you mean like this?
```
private void validateReplicationTypeRestoreSettings(IndexMetadata snapshottedMetadata, IndexMetadata updatedMetadata)
```

File: server/src/internalClusterTest/java/org/opensearch/indices/replication/SearchReplicaRestoreIT.java (hunk 6 above)
Comment (vinaykpud) [reply server/src/internalClusterTest/java/org/opensearch/indices/replication/SearchReplicaRestoreIT.java:225]: Okay

File: server/src/internalClusterTest/java/org/opensearch/indices/replication/SearchReplicaRestoreIT.java (hunk 8 above)
Comment (vinaykpud) [reply server/src/internalClusterTest/java/org/opensearch/indices/replication/SearchReplicaRestoreIT.java:48]: That makes sense, there are other 2 tests following this which has similar behaviour

File: server/src/internalClusterTest/java/org/opensearch/indices/replication/SearchReplicaRestoreIT.java (hunk 4 above)
Comment (vinaykpud) [reply server/src/internalClusterTest/java/org/opensearch/indices/replication/SearchReplicaRestoreIT.java:236]: Sure

File: server/src/internalClusterTest/java/org/opensearch/indices/replication/SearchReplicaRestoreIT.java (hunk 5 above)
Comment (vinaykpud) [reply server/src/internalClusterTest/java/org/opensearch/indices/replication/SearchReplicaRestoreIT.java:243]: Thats right

File: server/src/internalClusterTest/java/org/opensearch/indices/replication/SearchReplicaRestoreIT.java (hunk 7 above)
Comment (vinaykpud) [reply server/src/internalClusterTest/java/org/opensearch/indices/replication/SearchReplicaRestoreIT.java:228]: Sure

File: server/src/internalClusterTest/java/org/opensearch/indices/replication/SearchReplicaRestoreIT.java (hunk 3 above)
Comment (vinaykpud) [reply server/src/internalClusterTest/java/org/opensearch/indices/replication/SearchReplicaRestoreIT.java:255]: Done

File: server/src/internalClusterTest/java/org/opensearch/indices/replication/SearchReplicaRestoreIT.java [hunk 9]
```diff
@@ -0,0 +1,262 @@
...
+                Settings.builder()
+                    .put(IndexMetadata.SETTING_REPLICATION_TYPE, ReplicationType.DOCUMENT)
+                    .put(SETTING_NUMBER_OF_SEARCH_REPLICAS, 1)
+                    .build()
+            )
+        );
+        assertTrue(exception.getMessage().contains(getSnapshotExceptionMessage(ReplicationType.DOCUMENT, ReplicationType.DOCUMENT)));
+    }
+
+    public void testSearchReplicaRestore_WhenSnapshotOnDocRep_RestoreOnSegRep() throws Exception {
```
Comment (mch2) [inline server/src/internalClusterTest/java/org/opensearch/indices/replication/SearchReplicaRestoreIT.java:63]: We don't need this case here, this is covered by existing segrep restore tests.  Only the cases where search replicas are involved

File: server/src/internalClusterTest/java/org/opensearch/indices/replication/SearchReplicaRestoreIT.java [hunk 10]
```diff
@@ -0,0 +1,262 @@
...
+        ensureYellowAndNoInitializingShards(RESTORED_INDEX_NAME);
+        internalCluster().startDataOnlyNode();
+        ensureGreen(RESTORED_INDEX_NAME);
+        assertEquals(1, getNumberOfSearchReplicas(RESTORED_INDEX_NAME));
+
+        SearchResponse resp = client().prepareSearch(RESTORED_INDEX_NAME).setQuery(QueryBuilders.matchAllQuery()).get();
+        assertHitCount(resp, DOC_COUNT);
+    }
+
+    public void testSearchReplicaRestore_WhenSnapshotOnSegRep_RestoreOnDocRep() throws Exception {
```
Comment (mch2) [inline server/src/internalClusterTest/java/org/opensearch/indices/replication/SearchReplicaRestoreIT.java:104]: also don't need this case if the original index had no SR

File: CHANGELOG.md [hunk 11]
```diff
@@ -24,6 +24,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 - New `phone` & `phone-search` analyzer + tokenizer ([#15915](https://github.com/opensearch-project/OpenSearch/pull/15915))
 - Add _list/shards API as paginated alternate to _cat/shards ([#14641](https://github.com/opensearch-project/OpenSearch/pull/14641))
 - Latency and Memory allocation improvements to Multi Term Aggregation queries ([#14993](https://github.com/opensearch-project/OpenSearch/pull/14993))
+- Add support for restoring from snapshot with search replicas - ([#16111](https://github.com/opensearch-project/OpenSearch/pull/16111))
```
[the remark this answers is not in the document]
Comment (vinaykpud) [reply CHANGELOG.md:27]: Sure

File: server/src/internalClusterTest/java/org/opensearch/indices/replication/SearchReplicaRestoreIT.java (hunk 9 above)
Comment (vinaykpud) [reply server/src/internalClusterTest/java/org/opensearch/indices/replication/SearchReplicaRestoreIT.java:63]: Sure

File: server/src/internalClusterTest/java/org/opensearch/indices/replication/SearchReplicaRestoreIT.java (hunk 10 above)
Comment (vinaykpud) [reply server/src/internalClusterTest/java/org/opensearch/indices/replication/SearchReplicaRestoreIT.java:104]: Sure

File: CHANGELOG.md [hunk 12]
```diff
@@ -27,6 +27,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 - Latency and Memory allocation improvements to Multi Term Aggregation queries ([#14993](https://github.com/opensearch-project/OpenSearch/pull/14993))
 - Flat object field use IndexOrDocValuesQuery to optimize query ([#14383](https://github.com/opensearch-project/OpenSearch/issues/14383))
 - Add method to return dynamic SecureTransportParameters from SecureTransportSettingsProvider interface ([#16387](https://github.com/opensearch-project/OpenSearch/pull/16387)
```
Comment (mch2) [inline CHANGELOG.md:29]:
```suggestion
- Add method to return dynamic SecureTransportParameters from SecureTransportSettingsProvider interface ([#16387](https://github.com/opensearch-project/OpenSearch/pull/16387))
```

**Review by mch2 — APPROVED** [review]

**Comment by mch2** [comment]
> ❌ Gradle check result for [ca2fd04](https://github.com/opensearch-project/OpenSearch/commit/ca2fd046612c858ed405e9e2559796c0340d1f55): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/49888/)
> 
> Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green. Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

https://github.com/opensearch-project/OpenSearch/issues/15812

**Pull request merged**